test/lit/passes/multi-memory-lowering.wast

```diff
@@ -0,0 +1,214 @@
+;; RUN: wasm-opt %s --enable-multi-memories --multi-memory-lowering --enable-bulk-memory --enable-extended-const -S -o - | filecheck %s
+
+(module
+  (memory $memory1 1)
+  (memory $memory2 2)
+  (memory $memory3 3)
+  (data (memory $memory1) (i32.const 0) "a")
+  (data (memory $memory3) (i32.const 1) "123")
+  ;; CHECK:      (type $none_=>_i32 (func (result i32)))
+
+  ;; CHECK:      (type $i32_=>_i32 (func (param i32) (result i32)))
+
+  ;; CHECK:      (type $none_=>_none (func))
+
+  ;; CHECK:      (global $memory2_page_offset (mut i32) (i32.const 1))
+
+  ;; CHECK:      (global $memory3_page_offset (mut i32) (i32.const 3))
+
+  ;; CHECK:      (memory $combined_memory 6)
+
+  ;; CHECK:      (data (i32.const 0) "a")
+
+  ;; CHECK:      (data (i32.add
+  ;; CHECK-NEXT:  (global.get $memory3_page_offset)
+  ;; CHECK-NEXT:  (i32.const 1)
+  ;; CHECK-NEXT: ) "123")
+
+  ;; CHECK:      (func $loads
+  ;; CHECK-NEXT:  (drop
+  ;; CHECK-NEXT:   (i32.load
+  ;; CHECK-NEXT:    (i32.const 10)
+  ;; CHECK-NEXT:   )
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT:  (drop
+  ;; CHECK-NEXT:   (i32.load
+  ;; CHECK-NEXT:    (i32.add
+  ;; CHECK-NEXT:     (global.get $memory2_page_offset)
+  ;; CHECK-NEXT:     (i32.const 11)
+  ;; CHECK-NEXT:    )
+  ;; CHECK-NEXT:   )
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT:  (drop
+  ;; CHECK-NEXT:   (i32.load
+  ;; CHECK-NEXT:    (i32.add
+  ;; CHECK-NEXT:     (global.get $memory3_page_offset)
+  ;; CHECK-NEXT:     (i32.const 12)
+  ;; CHECK-NEXT:    )
+  ;; CHECK-NEXT:   )
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT: )
+  (func $loads
+  (drop
+   (i32.load $memory1
+    (i32.const 10)
+   )
+  )
+  (drop
+   (i32.load $memory2
+    (i32.const 11)
+   )
+  )
+  (drop
+   (i32.load $memory3
+    (i32.const 12)
+   )
+  )
+  )
+  ;; CHECK:      (func $stores
+  ;; CHECK-NEXT:  (i32.store
+  ;; CHECK-NEXT:   (i32.const 10)
+  ;; CHECK-NEXT:   (i32.const 115)
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT:  (i32.store
+  ;; CHECK-NEXT:   (i32.add
+  ;; CHECK-NEXT:    (global.get $memory2_page_offset)
+  ;; CHECK-NEXT:    (i32.const 11)
+  ;; CHECK-NEXT:   )
+  ;; CHECK-NEXT:   (i32.const 115)
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT:  (i32.store
+  ;; CHECK-NEXT:   (i32.add
+  ;; CHECK-NEXT:    (global.get $memory3_page_offset)
+  ;; CHECK-NEXT:    (i32.const 12)
+  ;; CHECK-NEXT:   )
+  ;; CHECK-NEXT:   (i32.const 115)
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT: )
+  (func $stores
+  (i32.store $memory1
+   (i32.const 10)
+   (i32.const 115)
+  )
+  (i32.store $memory2
+   (i32.const 11)
+   (i32.const 115)
+  )
+  (i32.store $memory3
+   (i32.const 12)
+   (i32.const 115)
+  )
+  )
+)
+
+;; CHECK:      (func $memory1_size (result i32)
+;; CHECK-NEXT:  (return
+;; CHECK-NEXT:   (global.get $memory2_page_offset)
+;; CHECK-NEXT:  )
+;; CHECK-NEXT: )
+
+;; CHECK:      (func $memory2_size (result i32)
+;; CHECK-NEXT:  (return
+;; CHECK-NEXT:   (i32.sub
+;; CHECK-NEXT:    (global.get $memory3_page_offset)
+;; CHECK-NEXT:    (global.get $memory2_page_offset)
```

Comment:
This looks wrong - `memory.size` returns a value in page size units, so I think this needs to be divided by the page size.

`memory.grow` also takes page sizes. I think the only place that needs the actual byte offsets are loads and stores.

Reply:
Spoke with Alon offline and explained that the (memoryName)_page_offset globals created for each memory are in page units. As a result, the load/store offsets need to change, with the page_offset being multiplied by the page size to get the byte offset that will be added to the existing load/store offset.

Given that we expect there to be far more load/store instructions than memory.grow/ memory.size, I’ll change the (memoryName)_page_offset globals to instead store the byte offset so that we’re able to just conduct a single add operation in each load/store instruction. For memory.grow/memory.size, I’ll divide the byte offset by the page size to get the page units needed for these instructions.